Update and expand docs for conda-pupa integration (#62)
conda-incubator/conda-pypi: docs/architecture.md

Reviewer: Let's move this file into a `developers/` section.

# Architecture

This page documents the technical architecture of `conda-pypi`, explaining
how it integrates with conda and the internal organization of its components.

## Plugin System Integration

`conda-pypi` is implemented as a conda plugin using conda's official plugin
architecture. This design enables seamless integration with conda's existing
workflows without requiring modifications to conda itself.

> The plugin registers several hooks with conda's plugin system. The

Reviewer: ```suggestion
The plugin registers several hooks with `conda`'s plugin system. The
```

subcommand hook adds the `conda pypi` subcommand to conda through
`conda_pypi.plugin.conda_subcommands()`, providing both `conda pypi install`
for installing PyPI packages with conversion and `conda pypi convert` for
converting PyPI packages without installing them.

The plugin also registers three post-command hooks that extend conda's
existing commands. The environment protection hook triggers after `install`,
`create`, `update`, and `remove` commands to automatically deploy

> `EXTERNALLY-MANAGED` files that prevent accidental pip usage. This is

Reviewer: ```suggestion
`EXTERNALLY-MANAGED` files that prevent accidental `pip` (or any other Python install tool) usage. This is
```

implemented through `ensure_target_env_has_externally_managed()`.

The explicit lockfile hook activates after `conda list --explicit` commands
to add PyPI package information as comments to explicit lockfiles. The
implementation in `_post_command_list_explicit()` appends `# pypi:` comment

> lines to lockfile output.

Author: I removed this section, which will be replaced by a future export plugin hook.

  - Environment path management
  - Package name normalization
  - `EXTERNALLY-MANAGED` file management
- **Design**: Consolidated utility functions to avoid code duplication

#### `mapping.py` - Package Name Mapping
- **Role**: Map PyPI package names to conda equivalents
- **Key Functions**:
  - `pypi_to_conda_name()` - Convert PyPI names to conda names
  - `conda_to_pypi_name()` - Reverse mapping
  - `get_grayskull_mapping()` - Load automated mapping data
- **Data Sources**: Uses grayskull project mappings for automated name
  translation

#### `vcs.py` - Version Control Integration
- **Role**: Handle VCS URLs for editable installations
- **Key Components**:
  - `VCSHandler` - Support for git, hg, svn, bzr
  - `VCSInfo` - VCS URL parsing and metadata
- **Capabilities**:
  - Parse VCS URLs with branches, refs, subdirectories
  - Clone repositories for editable installs
  - Support multiple VCS systems

#### `main.py` - Environment Management
- **Role**: Environment validation and lockfile functionality
- **Key Functions**:
  - `validate_target_env()` - Check environment requirements
  - `pypi_lines_for_explicit_lockfile()` - Generate PyPI lockfile lines

>   - Environment compatibility checking

Author: This section was removed.

#### `exceptions.py` - Error Handling
- **Role**: Define conda-pypi specific exceptions
- **Components**: Custom exception classes that integrate with conda's error
  system

## Data Flow Architecture

### Installation Flow

```
conda pypi install package
         ↓
CLI Argument Parsing
         ↓
Environment Validation
         ↓
Package Classification
         ↓

>     VCS/Editable? ----Yes---→ Direct pip install

Reviewer: We don't install things with conda in editable installs?

         ↓
Fetch from PyPI
         ↓
Download Wheels
         ↓
Convert to .conda
         ↓
Save to Output Directory
```

### Plugin Hook Flow

```
conda command executed
         ↓
Post-command hook?
    ↓        ↓           ↓
list      install/    install/create/
--explicit  create     update/remove
    ↓        ↓           ↓
Add PyPI  Process    Deploy
lines     PyPI       EXTERNALLY-
    ↓     lines      MANAGED
Output to   ↓           ↓
stdout   Install    Create marker
         PyPI       files
         packages
```

## Key Design Principles

The architecture of `conda-pypi` is built around several key design
principles that ensure effective integration between conda and PyPI
ecosystems.

Conda-native integration is achieved by using conda's official plugin
system and leveraging conda's existing infrastructure including solvers,
channels, and metadata systems. This approach maintains full compatibility
with existing conda workflows.

The system emphasizes modularity and separation of concerns with clear
separation between CLI, core logic, and specialized functionality. Each module
has a focused responsibility with minimal coupling between modules, making the

> codebase easier to maintain and extend.

Author: Removed.

The intelligent hybrid approach ensures that explicit packages always come
from PyPI to respect user intent, while dependencies prefer conda channels
for ecosystem compatibility. The system falls back to PyPI conversion only
when needed.

> Extensibility is built into the design through the plugin architecture that
> allows easy extension, modular design that supports new conversion engines,
> and VCS system support that can be easily extended to new version control
> systems.

Reviewer: Also superfluous, I feel.
```suggestion
```
